[SPARK-46094] Support Executor JVM Profiling

parthchandra opened the pull request:
### What changes were proposed in this pull request?
This adds support for the async profiler to Spark


### Why are the changes needed?

Profiling of JVM applications on a cluster is cumbersome and it can be complicated to save the output of the profiler especially if the cluster is on K8s where the executor pods are removed and any files saved to the local file system become inaccessible. This feature makes it simple to turn profiling on/off, includes the jar/binaries needed for profiling,  and makes it simple to save output to an HDFS location.


### Does this PR introduce _any_ user-facing change?
This PR introduces three new configuration parameters. These are described in the documentation.



### How was this patch tested?
Tested manually on EKS

### Was this patch authored or co-authored using generative AI tooling?
No

parthchandra committed:
[SPARK-46094][CONNECT] Add support for code profiling executors

This adds support for the async profiler to Spark

Profiling of JVM applications on a cluster is cumbersome and it can be complicated to save the output of the profiler especially if the cluster is on K8s where the executor pods are removed and any files saved to the local file system become inaccessible.
This feature makes it simple to turn profiling on/off, includes the jar/binaries needed for profiling,  and makes it simple to save output to an HDFS location.

This PR introduces three new configuration parameters. These are described in the documentation.

HyukjinKwon commented:
how do you use it? would be great if it contains the example, how to run, etc.

parthchandra commented:
> how do you use it? would be great if it contains the example, how to run, etc.

There's a README - connector/profiler/README.md. I can add more details if you think this is not enough.

parthchandra commented:
There's a whole slew of errors like the following while building documentation - 
```
[error] /__w/spark/spark/Loading source file /__w/spark/spark/common/sketch/src/main/java/org/apache/spark/util/sketch/BitArray.java...
[error] Loading source file /__w/spark/spark/common/sketch/src/main/java/org/apache/spark/util/sketch/BloomFilter.java...
[error] Loading source file /__w/spark/spark/common/sketch/src/main/java/org/apache/spark/util/sketch/BloomFilterImpl.java...
[error] Loading source file /__w/spark/spark/common/sketch/src/main/java/org/apache/spark/util/sketch/CountMinSketch.java...
[error] Loading source file /__w/spark/spark/common/sketch/src/main/java/org/apache/spark/util/sketch/CountMinSketchImpl.java...
```

How does one fix this?

dongjoon-hyun reviewed (commented):
Thank you for making a PR, @parthchandra .

dongjoon-hyun commented on connector/profiler/README.md:
````diff
@@ -0,0 +1,86 @@
+
+To build 
+``` 
+  ./build/mvn clean package -P code-profiler
````
In addition, please add `-DskipTests` in order to be more copy-and-paste friendly.

LuciferYang replied:
In Spark, it is more customary to write as `-Pcode-profiler`

dongjoon-hyun commented on connector/profiler/README.md:
```diff
@@ -0,0 +1,86 @@
+
+## Executor Code Profiling
+
+The spark profiler module enables code profiling of executors in cluster mode based on the the [async profiler](https://github.com/async-profiler/async-profiler/blob/master/README.md), a low overhead sampling profiler. This allows a Spark application to capture CPU and memory profiles for application running on a cluster which can later be analyzed for performance issues. The profiler captures [J ...
```
For Java 8 reference link, it looks inappropriate because Apache Spark 4.0.0 dropped all Java versions less than 16.
Do you think we can have Java 17+ link, @parthchandra ?

dongjoon-hyun commented on connector/profiler/README.md:
```diff
@@ -0,0 +1,86 @@
+
+The profiler writes the jfr files to the executor's working directory in the executor's local file system and the files can grow to be large so it is advisable that the executor machines have adequate storage. The profiler can be configured to copy the jfr files to a hdfs location before the executor shuts down.
+
+Code profiling is currently only supported for
```
Just a question. Why not Windows?

dongjoon-hyun commented on connector/profiler/README.md:
````diff
@@ -0,0 +1,86 @@
+To get maximum profiling information set the following jvm options for the executor -
+
+```
+    -XX:+UnlockDiagnosticVMOptions -XX:+DebugNonSafepoints -XX:+PreserveFramePointer
````
Let's remove the trailing space.

dongjoon-hyun commented on connector/profiler/README.md:
```diff
@@ -0,0 +1,86 @@
+*   Linux (musl, x64)
+*   MacOS
+
+To get maximum profiling information set the following jvm options for the executor -
```
`-` -> `:`

dongjoon-hyun commented on connector/profiler/README.md:
```diff
@@ -0,0 +1,86 @@
+  <td><code>spark.executor.profiling.enabled</code></td>
+  <td>
+    <code>false</code>
+  </td>
```
Let's unify the style. Like 74 line, one-liner (`<td><code>false</code></td>`) is better.

dongjoon-hyun commented on connector/profiler/README.md:
```diff
@@ -0,0 +1,86 @@
+</tr>
+<tr>
+  <td><code>spark.executor.profiling.outputDir</code></td>
+  <td></td>
```
Please use `(none)`.
https://github.com/apache/spark/blob/ec71e223fc4f2d1e3a444d5c4fff8b281dc7b58c/docs/configuration.md?plain=1#L354

dongjoon-hyun commented on connector/profiler/README.md:
```diff
@@ -0,0 +1,86 @@
+  <td><code>spark.executor.profiling.outputDir</code></td>
+  <td></td>
+  <td>
+      An hdfs compatible path to which the profiler's output files are copied. The output files will be written as <i>outputDir/application_id/profile-appname-exec-executor_id.jfr</i> <br/>
```
`hdfs` -> `HDFS`

dongjoon-hyun commented on connector/profiler/README.md:
```diff
@@ -0,0 +1,86 @@
+  <td></td>
+  <td>
+      An hdfs compatible path to which the profiler's output files are copied. The output files will be written as <i>outputDir/application_id/profile-appname-exec-executor_id.jfr</i> <br/>
+      If no outputDir is specified then the files are not copied over. 
```
Please add a warning about Out-Of-Disk situation because K8s is very strict about the disk usage unlike YARN or Standalone clusters.

dongjoon-hyun commented on connector/profiler/README.md:
```diff
@@ -0,0 +1,86 @@
+  <td>4.0.0</td>
+</tr>
+<tr>
+  <td><code>spark.executor.profiling.outputDir</code></td>
```
Since this is remote HDFS-compatible location, let's follow our convention like the following.

https://github.com/apache/spark/blob/ec71e223fc4f2d1e3a444d5c4fff8b281dc7b58c/docs/configuration.md?plain=1#L442

In short, please rename `outputDir` to `dfsDir`.
```
- spark.executor.profiling.outputDir
+ spark.executor.profiling.dfsDir
```

dongjoon-hyun commented on connector/profiler/README.md:
```diff
@@ -0,0 +1,86 @@
+</tr>
+</table>
+
+### Kubernetes
```
Thank you for adding this.

dongjoon-hyun commented on connector/profiler/README.md:
````diff
@@ -0,0 +1,86 @@
+### Kubernetes
+On Kubernetes, spark will try to shut down the executor pods while the profiler files are still being saved. To prevent this set 
+```
+  spark.kubernetes.executor.deleteOnTermination=false
````
Remove the trailing spaces.

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/ExecutorCodeProfiler.scala:
```diff
@@ -0,0 +1,173 @@
+
+/**
+ * A class that enables the async code profiler
+ *
```
nit. Remove the empty line.

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/ExecutorCodeProfiler.scala:
```diff
@@ -0,0 +1,173 @@
+ * A class that enables the async code profiler
+ *
+ */
+private[spark] class ExecutorCodeProfiler(conf: SparkConf, executorId: String) extends Logging {
```
Spark Executors can have Java, Puython, and R runtimes. Given that, `Code` is a little vague term.

- I'd like to propose to rename `ExecutorCodeProfiler` to `ExecutorJVMProfiler`.
- Otherwise, at least, please document it clearly that this is JVM-only feature.

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/ExecutorCodeProfiler.scala:
```diff
@@ -0,0 +1,173 @@
+  private val resumecmd = s"resume,$profilerOptions,file=$profilerLocalDir/profile.jfr"
+
+  private val UPLOAD_SIZE = 8 * 1024 * 1024 // 8 MB
+  private val WRITE_INTERVAL = 30 // seconds
```
Why this is a magic number instead of a configuration?

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/ExecutorCodeProfiler.scala:
```diff
@@ -0,0 +1,173 @@
+  private var writing: Boolean = false
+
+  val profiler: AsyncProfiler = if (enableProfiler) {
+    if (AsyncProfilerLoader.isSupported) {
```
Could you give me a link for this method, please?

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/ExecutorProfilerPlugin.scala:
```diff
@@ -0,0 +1,75 @@
+
+/**
+ * Spark plugin to do code profiling of executors
+ *
```
nit. Remove this line.

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/ExecutorProfilerPlugin.scala:
```diff
@@ -0,0 +1,75 @@
+  }
+
+  override def shutdown(): Unit = {
+
```
nit. Remove this line.

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/ExecutorProfilerPlugin.scala:
```diff
@@ -0,0 +1,75 @@
+    }
+  }
+
+
```
nit. Remove the above two lines.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -799,6 +799,44 @@ package object config {
+      .checkValue(v => v >= 0.0 && v < 1.0,
+        "Fraction of executors to profile must be in [0,1)")
+      .createWithDefault(0.1)
+
```
If you don't mind, please move this into a seperate `package` like Kafka module under `connector`.

https://github.com/apache/spark/blob/7a0d0411aa02e7e1b6beb393966ada2c54c09870/connector/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/package.scala#L25-L34

dongjoon-hyun replied:
For YARN and K8s, we have `Config.scala`.

https://github.com/apache/spark/blob/7a0d0411aa02e7e1b6beb393966ada2c54c09870/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala#L26

https://github.com/apache/spark/blob/7a0d0411aa02e7e1b6beb393966ada2c54c09870/resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/config.scala#L27

dongjoon-hyun reviewed (commented):
It's a great addition, @parthchandra . Thank you so much! I finished my first review with a few comments. Other designs look good to me.

dongjoon-hyun commented:
Oh, one more thing, I removed `[CONNECT]` from the title because `[CONNECT]` is for `Spark Connect`.

LuciferYang commented:
> There's a whole slew of errors like the following while building documentation -
> 
> ```
> [error] /__w/spark/spark/Loading source file /__w/spark/spark/common/sketch/src/main/java/org/apache/spark/util/sketch/BitArray.java...
> [error] Loading source file /__w/spark/spark/common/sketch/src/main/java/org/apache/spark/util/sketch/BloomFilter.java...
> [error] Loading source file /__w/spark/spark/common/sketch/src/main/java/org/apache/spark/util/sketch/BloomFilterImpl.java...
> [error] Loading source file /__w/spark/spark/common/sketch/src/main/java/org/apache/spark/util/sketch/CountMinSketch.java...
> [error] Loading source file /__w/spark/spark/common/sketch/src/main/java/org/apache/spark/util/sketch/CountMinSketchImpl.java...
> ```
> 
> How does one fix this?

The actual error is

```
Warning, treated as error:
don't know which module to import for autodocumenting 'apply_batch' (try placing a "module" or "currentmodule" directive in the document, or giving an explicit module name)
make: *** [Makefile:35: html] Error 2
                    ------------------------------------------------
      Jekyll 4.3.2   Please append `--trace` to the `build` command 
                     for any additional information or backtrace. 
                    ------------------------------------------------
/__w/spark/spark/docs/_plugins/copy_api_dirs.rb:130:in `<top (required)>': Python doc generation failed (RuntimeError)
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.3.2/lib/jekyll/external.rb:57:in `require'
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.3.2/lib/jekyll/external.rb:57:in `block in require_with_graceful_fail'
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.3.2/lib/jekyll/external.rb:55:in `each'
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.3.2/lib/jekyll/external.rb:55:in `require_with_graceful_fail'
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.3.2/lib/jekyll/plugin_manager.rb:96:in `block in require_plugin_files'
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.3.2/lib/jekyll/plugin_manager.rb:94:in `each'
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.3.2/lib/jekyll/plugin_manager.rb:94:in `require_plugin_files'
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.3.2/lib/jekyll/plugin_manager.rb:21:in `conscientious_require'
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.3.2/lib/jekyll/site.rb:131:in `setup'
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.3.2/lib/jekyll/site.rb:36:in `initialize'
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.3.2/lib/jekyll/commands/build.rb:30:in `new'
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.3.2/lib/jekyll/commands/build.rb:30:in `process'
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.3.2/lib/jekyll/command.rb:91:in `block in process_with_graceful_fail'
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.3.2/lib/jekyll/command.rb:91:in `each'
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.3.2/lib/jekyll/command.rb:91:in `process_with_graceful_fail'
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.3.2/lib/jekyll/commands/build.rb:18:in `block (2 levels) in init_with_program'
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `block in execute'
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `each'
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `execute'
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/mercenary-0.4.0/lib/mercenary/program.rb:44:in `go'
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/mercenary-0.4.0/lib/mercenary.rb:21:in `program'
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.3.2/exe/jekyll:15:in `<top (required)>'
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/bin/jekyll:23:in `load'
	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/bin/jekyll:23:in `<main>'
```

parthchandra committed:
Address review comments

parthchandra commented:
@dongjoon-hyun addressed your comments. (Sorry it took a while to address the changes)
@HyukjinKwon, I've added an explicit usage example to the README

dongjoon-hyun commented:
Thank you for updates, @parthchandra .

dongjoon-hyun commented on connector/profiler/README.md:
```diff
@@ -0,0 +1,109 @@
+  -c spark.executor.extraJavaOptions="-XX:+UnlockDiagnosticVMOptions -XX:+DebugNonSafepoints -XX:+PreserveFramePointer" \
+  -c spark.plugins=org.apache.spark.executor.profiler.ExecutorProfilerPlugin \
+  -c spark.executor.profiling.enabled=true \
+  -c spark.executor.profiling.outputDir=s3a://my-bucket/spark/profiles/  \
```
`spark.executor.profiling.dfsDir`?

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/profiler/package.scala:
```diff
@@ -0,0 +1,68 @@
+
+  private[profiler] val EXECUTOR_CODE_PROFILING_OUTPUT_DIR =
+    ConfigBuilder("spark.executor.profiling.dfsDir")
+      .doc("HDFS compatible file-system  path to where the profiler will write output jfr files.")
```
`file-system  path` -> `file-system path`

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/profiler/package.scala:
```diff
@@ -0,0 +1,68 @@
+
+  private[profiler] val EXECUTOR_CODE_PROFILING_OPTIONS =
+    ConfigBuilder("spark.executor.profiling.options")
+      .doc("Options to pass on to the async profiler.")
```
Could you add some pointer for async profiler values?

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/profiler/package.scala:
```diff
@@ -0,0 +1,68 @@
+      .version("4.0.0")
+      .doubleConf
+      .checkValue(v => v >= 0.0 && v < 1.0,
+        "Fraction of executors to profile must be in [0,1)")
```
Why do we exclude 100%?

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/profiler/package.scala:
```diff
@@ -0,0 +1,68 @@
+    ConfigBuilder("spark.executor.profiling.writeInterval")
+      .doc("Time interval in seconds after which the profiler output will be synced to dfs")
+      .version("4.0.0")
+      .intConf
```
Please use `.timeConf(TimeUnit.SECONDS)` instead of `intConf`.

dongjoon-hyun replied:
In addition, shall we add `checkValue`? For example, the following?
```scala
.checkValue(_ >= 0, "Write interval should be non-negative")
```

Is it positive or non-negative? Please specify our limit here.

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/profiler/ExecutorProfilerPlugin.scala:
```diff
@@ -0,0 +1,70 @@
+  override def executorPlugin(): ExecutorPlugin = new CodeProfilerExecutorPlugin
+}
+
+class CodeProfilerExecutorPlugin extends ExecutorPlugin with Logging {
```
`CodeProfilerExecutorPlugin` -> `JVMProfilerExecutorPlugin`

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/profiler/ExecutorProfilerPlugin.scala:
```diff
@@ -0,0 +1,70 @@
+
+
+/**
+ * Spark plugin to do code profiling of executors
```
`code` -> `JVM`

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/profiler/ExecutorJVMProfiler.scala:
```diff
@@ -0,0 +1,171 @@
+        inputStream.close()
+        outputStream.close()
+      } catch {
+        case e: Exception =>
```
Please use a specific exception instead of `Exception`.

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/profiler/ExecutorJVMProfiler.scala:
```diff
@@ -0,0 +1,171 @@
+        outputStream.close()
+      } catch {
+        case e: Exception =>
+          logError("Exception in completing profiler output", e)
```
Could you add a descriptive message about what happens if we log and ignore this exception?

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/profiler/ExecutorJVMProfiler.scala:
```diff
@@ -0,0 +1,171 @@
+  private var running = false
+  private val enableProfiler = conf.get(EXECUTOR_CODE_PROFILING_ENABLED)
+  private val profilerOptions = conf.get(EXECUTOR_CODE_PROFILING_OPTIONS)
+  private val profilerOutputDir = conf.get(EXECUTOR_CODE_PROFILING_OUTPUT_DIR)
```
After renaming `EXECUTOR_CODE_PROFILING_OUTPUT_DIR`, we need to rename this variable too.

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/profiler/ExecutorJVMProfiler.scala:
```diff
@@ -0,0 +1,171 @@
+  private var threadpool: ScheduledExecutorService = _
+  private var writing: Boolean = false
+
+  val profiler: AsyncProfiler = if (enableProfiler) {
```
We prefer a Scala-way, `Option`, for nullable variable.

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/profiler/ExecutorJVMProfiler.scala:
```diff
@@ -0,0 +1,171 @@
+  }
+
+  def start(): Unit = {
+    if (profiler != null && !running) {
```
Please change this after we change `profiler` into `Option[AsyncProfiler]`

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/profiler/ExecutorJVMProfiler.scala:
```diff
@@ -0,0 +1,171 @@
+      try {
+        profiler.execute(startcmd)
+      } catch {
+        case e: Exception =>
```
Please use a more specific exception.

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/profiler/ExecutorJVMProfiler.scala:
```diff
@@ -0,0 +1,171 @@
+      } catch {
+        case e: Exception =>
+          logWarning("Executor code profiling aborted due to exception: ", e)
+          return
```
May I ask why we prevent the exception propagation? I believe it's better instead of ignoring like this.

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/profiler/ExecutorJVMProfiler.scala:
```diff
@@ -0,0 +1,171 @@
+    startWriting()
+  }
+
+  /** Stops the profiling and saves output to hdfs location. */
```
Since `hdfs` could be misleading, `dfs`?

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/profiler/ExecutorJVMProfiler.scala:
```diff
@@ -0,0 +1,171 @@
+  def stop(): Unit = {
+    if (profiler != null && running) {
+      profiler.execute(stopcmd)
+      logInfo("Code profiler stopped")
```
`Code` -> `JVM`?

dongjoon-hyun reviewed (commented):
Thank you for updating, @parthchandra . I finished the second review.

I believe we are close. Could you address the comments?

parthchandra committed:
Address more review comments.

parthchandra commented:
@dongjoon-hyun I've addressed your comments. I also made some minor fixes to address some errors I saw while running locally.

parthchandra commented:
> A couple of queries:
> 
> a) IIRC async profiler does not have a way to associate the java thread with the underlying linux thread - and so from a future evolution point of view, we cannot corelate this with a task's stack traces - is this still the case ? If yes, given work being done as part of [SPARK-45151](https://issues.apache.org/jira/browse/SPARK-45151), is there a way to make this work with whatever @yaooqinn added support for ? ([honest-profiler](https://github.com/jvm-profiling-tools/honest-profiler) has support for this, but unfortunately it does not have maven artifacts yet :-( )

I believe async_profiler uses the JVM's internal `AsyncGetCallTrace` method (https://github.com/async-profiler/async-profiler?tab=readme-ov-file#cpu-profiling) to walk the stack for a thread including native calls made by the JVM. AFAIK, honest-profiler does the same. There is also the `AsyncGetStackTrace` API in the works https ://openjdk.org/jeps/435 (not sure if this is available yet). At some point we can investigate how to leverage the new API to add native calls to the work done in SPARK-45151.

> b) This PR is creating one file per executor on hdfs - which would end up being very expensive as number of executors per application and number of applications increase. How are we looking to minimize this cost ?

There is a configuration `spark.executor.profiling.fraction` to limit the number of executors to be profiled (default is 0.1). One can also specify profiler options to limit the sampling frequecy `spark.executor.profiling.options=...,interval=10ms,...` which can limit the size of the jfr file created (useful for very long running jobs).

mridulm commented:
If async profiler does not allow us to map the native thread to its java thread (please validate this) - we cannot map stack traces to the corresponding task threads - and so limits usability of this integration in spark.
If this limitation exists, we should explore alternatives which might support it and is easy to integrate (honest profiler supports it but is not easy to integrate iirc).


Simply dumping per executor flamegraphs or stack traces has limited utility (and can be done today) unless we have a path to integrating with SPARK-45151 IMO.

Thoughts @dongjoon-hyun ?

parthchandra commented:
> If async profiler does not allow us to map the native thread to its java thread (please validate this) - we cannot map stack traces to the corresponding task threads - and so limits usability of this integration in spark.

`AsyncGetCallTrace` is used precisely to map calls in the native thread to calls in the java thread. 
Not sure exactly what you are looking for here. Are you looking to profile individual tasks? It certainly can be done, but would require some changes similar to SPARK-45151 and some additional work if you want the profile available thru the UI. Or are you looking to enhance SPARK-45151 and get a stack trace that includes native calls? This is a little harder via async_profiler since there is no API to get a snapshot.
Note that getting a profile needs to be collected over a period of time and so is different from getting a snapshot as SPARK-45151 is doing.

> Simply dumping per executor flamegraphs or stack traces has limited utility (and can be done today).
I would suggest that this PR makes it trivially simple to profile with no setup required. On K8s, with ephemeral storage, it is not a simple task to dump a profile to disk and get it off the pod before the pod is destroyed (it was in fact the original motivation behind doing this).

mridulm commented:
> `AsyncGetCallTrace` is used precisely to map calls in the native thread to calls in the java thread. Not sure exactly what you are looking for here. Are you looking to profile individual tasks? It certainly can be done, but would require some changes similar to [SPARK-45151](https://issues.apache.org/jira/browse/SPARK-45151) and some additional work if you want the profile available thru the UI. Or are you looking to enhance [SPARK-45151](https://issues.apache.org/jira/browse/SPARK-45151) and get a stack trace that includes native calls? This is a little harder via async_profiler since there is no API to get a snapshot. Note that getting a profile needs to be collected over a period of time and so is different from getting a snapshot as [SPARK-45151](https://issues.apache.org/jira/browse/SPARK-45151) is doing.


There is a difference between native thread id's and java thread ids.
Given the async profiler output, can we map it to the corresponding task (given task's java thread id) ?
My understanding is currently no - but if I am missing something, do let me know.

Assuming no, this means the stack traces generated are for all threads in the executor jvm - and so does not allow us to get stack traces and/or flamegraphs for a particular task, tasks of a stage, etc.

If yes, this would be very useful - and will allow for future evolution as part of SPARK-44893 [1].


> 
> > Simply dumping per executor flamegraphs or stack traces has limited utility (and can be done today).
> 
> I would suggest that this PR makes it trivially simple to profile with no setup required. On K8s, with ephemeral storage, it is not a simple task to dump a profile to disk and get it off the pod before the pod is destroyed (it was in fact the original motivation behind doing this).

I am not seeing a lot of value in including this into Apache Spark itself - plugin api is public, and users can leverage it to do precisely what the PR is proposing.
On other hand, if the PR is integrating well with SPARK-44893 [1] - and/or there is a path to leveraging it in that work, it would be more useful.

I am not exactly -1 on this @dongjoon-hyun , but I am not seeing a lot of value in it: will let you make the call.


[1] This is the jira I was trying to paste, but github mobile messed it up - and ended up referencing a subtask !

parthchandra commented:
> There is a difference between native thread id's and java thread ids. Given the async profiler output, can we map it to the corresponding task (given task's java thread id) ? My understanding is currently no - but if I am missing something, do let me know.

Yes we can map the stack traces to the java thread. Here's how it looks (this is in intellij's profiler window)
<img width="2046" alt="Screenshot 2023-12-20 at 9 46 49 AM" src="https://github.com/apache/spark/assets/6529136/e87d4b9d-4e00-4a89-aa50-c107399359bc">

> Assuming no, this means the stack traces generated are for all threads in the executor jvm - and so does not allow us to get stack traces and/or flamegraphs for a particular task, tasks of a stage, etc.

We can get individual threads and even filter to profile a single thread. This PR specifically profiles every thread in the executor. 

> If yes, this would be very useful - and will allow for future evolution as part of [SPARK-44893](https://issues.apache.org/jira/browse/SPARK-44893) [1].

Ah, this JIRA makes it clearer. We can leverage the async-profiler to provide the features not yet implemented in  [SPARK-45209](https://issues.apache.org/jira/browse/SPARK-45209). The current implementation uses a simple snapshot of the task stack traces which can be enhanced by using the async-profiler to get accurate profiling.  
 
> I am not seeing a lot of value in including this into Apache Spark itself - plugin api is public, and users can leverage it to do precisely what the PR is proposing. On other hand, if the PR is integrating well with [SPARK-44893](https://issues.apache.org/jira/browse/SPARK-44893) [1] - and/or there is a path to leveraging it in that work, it would be more useful.

I think we can certainly leverage this work. This PR by itself does not have the APIs needed to enhance SPARK-45209. It would probably need to be a separate PR because it may need changes to the UI implementation. We can either get a flamegraph (covering a period of time for a task) or collapsed call traces from which a flamegraph can be produced and the choice will affect the UI.

> I am not -1 on this @dongjoon-hyun , but I am not seeing a lot of value in it: will let you make the call (also because I am on vacation, dont have my desktop handy to investigate in detail :) ).

🙏🏾

mridulm commented:
That sounds promising !
What is unclear to me is how we are going to do the mapping without something which ends up introducing safe point bias (essentially, cost of this operation) ... 
For example, if the native to java thread mapping requires mxbean.getThreadInfo and/or similar approaches, it becomes fairly expensive.

Essentially what I am trying to make sure is - given `(native-thread-id -> timestamp -> stack_dumps+)*`, can we identify the `native-thread -> java-thread-id` ?
If yes, we can build the `java-thread-id -> task-id` in spark, and essentially get to `(task-id -> stack_dumps+)*` for all (most ?) tasks.

When we built [Safari](https://blog.cloudera.com/demystifying-spark-jobs-to-optimize-for-cost-and-performance/), this is what ended up being extremely powerful for understanding application performance - per-task stack dumps, correlated across all tasks for a stage: allowing us to understand what the stack dump for a particular stage is, what the difference between 'expensive' tasks in a stage vs average task is, etc.

parthchandra commented:
That was pretty cool stuff you did in Safari!
I think we may not have to do too much work ourselves.
The way I see it, async_profiler is doing the mapping of the java threads and stack traces already for us (and we know that both async_profiler and honest_profiler avoid the safepoint bias problem so this is as good as it gets). In addition there is a filter API to filter on one or more threads so async_profiler collects events only for the given thread(s). The API for filtering takes a `java.lang.Thread` as input.
The way I see it potentially working is: when a user asks to profile a task, we start profiling for only the task's thread similar to the way a task stack trace is being done today. Then we ship over the the collected data and display it. 
I'll have to play around with this though. There might be some gotchas in profiling multiple threads simultaneously and/or some APIs might be private making it

dongjoon-hyun commented:
Sorry for being away, @mridulm and @parthchandra . I've been traveling in South Korea since 14th December. I'll catch up the discussion and will revisit this PR on January. Thank you!

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/profiler/ExecutorJVMProfiler.scala:
```diff
@@ -0,0 +1,183 @@
+          running = true
+          startWriting()
+        }
+        )
```
nit. Let's merge line 71 into 70.

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/profiler/ExecutorJVMProfiler.scala:
```diff
@@ -0,0 +1,183 @@
+        }
+        )
+      } catch {
+        case e@(_: IllegalArgumentException | _: IllegalStateException | _: IOException) =>
```
nit. Add proper spaces?
```scala
- case e@(
+ case e @ (
```

parthchandra replied:
Done

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/profiler/ExecutorJVMProfiler.scala:
```diff
@@ -0,0 +1,183 @@
+        threadpool.scheduleWithFixedDelay(new Runnable() {
+          override def run(): Unit = writeChunk(false)
+        }, writeInterval, writeInterval,
+          TimeUnit.SECONDS)
```
The indentation of the above four lines looks weird to me. Could you revise?

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/profiler/ExecutorJVMProfiler.scala:
```diff
@@ -0,0 +1,183 @@
+      }
+    } catch {
+      case e: IOException => logError("Exception occurred while writing some profiler output: ", e)
+      case e@(_: IllegalArgumentException | _: IllegalStateException) =>
```
ditto. `e@(` -> `e @ (`

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/profiler/ExecutorJVMProfiler.scala:
```diff
@@ -0,0 +1,183 @@
+      case e@(_: IllegalArgumentException | _: IllegalStateException) =>
+        logError("Some profiler output not written." +
+          " Exception occurred in profiler native code: ", e)
+      case e: Exception => logError("Some profiler output not written. Unexpected exception: ", e)
```
I'm wondering what we can get in this case because the flag `writing` is still `true`. Do we need to keep `writing` because we need to invoke `finishWriting`? However, it seems that we cannot invoke `inputStream.close()` and `outputStream.close()` eventually.

dongjoon-hyun replied:
Got it. Since we have the error log, it might be okay for now.

dongjoon-hyun commented on connector/profiler/src/main/scala/org/apache/spark/executor/profiler/ExecutorJVMProfiler.scala:
```diff
@@ -0,0 +1,183 @@
+        outputStream.close()
+      } catch {
+        case _: InterruptedException => Thread.currentThread().interrupt()
+        case e: IOException =>
```
Where does this come from when `writeChunk` swallows all exceptions with `case e: Exception`?

dongjoon-hyun commented:
Thank you all. Especially, @mridulm for many ideas and advices on this PR reviews.
- Ya, as you mentioned, `honest-profiler` has been abandoned about two years unfortunately. We should not depend on it.
- Like `Safari` (2019), there has been many profiler approaches outside the community because it's valuable. However, IIUC, it's not easily available to the Apache Spark users in these days.
- I believe the value of this PR is capable of providing `Apache Spark 4.0.0` users the foundation of easy pre-defined profiling with additional help like `spark.executor.profiling.fraction`.
    - To be clear, we don't recommend users to enable this for all executors of all jobs.
- Lastly, this plugin way is non-intrusive and has no maintain cost because it's based on the Spark standard. For the proposed integrated and detailed analysis, I believe we can achieve them in the future in a way because we agreed lots of values of that integrated analysis.

I hope we can merge this as a part of Apache Spark 4.0.0.

parthchandra committed:
Address more review comments

dongjoon-hyun reviewed (approved):
+1, LGTM. (Pending CIs)

dongjoon-hyun commented:
Merged to master for Apache Spark 4.0.0.
Thank you, @parthchandra and all!

pan3793 commented on connector/profiler/src/main/scala/org/apache/spark/executor/profiler/package.scala:
```diff
@@ -0,0 +1,71 @@
+      .createWithDefault(false)
+
+  private[profiler] val EXECUTOR_PROFILING_DFS_DIR =
+    ConfigBuilder("spark.executor.profiling.dfsDir")
```
@dongjoon-hyun, I'm also investigating adding profiling support for the driver, should I fork all configurations, or simply treat the driver as a special executor and resue those configurations?

dongjoon-hyun replied:
Well, @pan3793 , could you use a new JIRA for your new goal, please?

Although I understand why you adds a comment here, it doesn't look like a good practice to me. Why do we add a comment about driver-related suggestion to the irrelevant executor-related PR like `Spark Executor JVM Profiling`?
